File: src/GUI/objects/InfoNode/TaskInfo.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package GUI.objects.InfoNode;

import javafx.concurrent.Task;
import javafx.scene.control.Labeled;
import javafx.scene.control.ProgressIndicator;

/**
 *  Provides information about the task and its progress.
 */
public class TaskInfo implements InfoNode<Task> {
    
    public final Labeled title;
    public final Labeled message;
    public final ProgressIndicator progressIndicator;

    /**
     * @param title title label. Use null if none.
     * @param message message label. Use null if none.
     * @param pi progress indicator. Use null if none.
     */
    public TaskInfo(Labeled title, Labeled message, ProgressIndicator pi) {
        this.title = title;
        this.message = message;
        this.progressIndicator = pi;
    }
    
    /** {@inheritDoc} */
    @Override
    public void setVisible(boolean v) {
        if(title!=null) title.setVisible(v);
        if(message!=null) message.setVisible(v);
        if(progressIndicator!=null) progressIndicator.setVisible(v);
    }
    
    /** {@inheritDoc} */
    @Override
    public void bind(Task t) {
        unbind();
        if(progressIndicator!=null) progressIndicator.progressProperty().bind(t.progressProperty());
        if(title!=null) title.textProperty().bind(t.titleProperty());
        if(message!=null) message.textProperty().bind(t.messageProperty());
    }
    
    /** {@inheritDoc} */
    @Override
    public void unbind() {
        if(progressIndicator!=null) progressIndicator.progressProperty().unbind();
        if(title!=null) title.textProperty().unbind();
        if(message!=null) message.textProperty().unbind();
    }
    
}

File: src/sp/it/pl/gui/nodeinfo/TaskInfo.kt
package sp.it.pl.gui.nodeinfo

import javafx.concurrent.Task
import javafx.concurrent.Worker.State.READY
import javafx.concurrent.Worker.State.SCHEDULED
import javafx.scene.control.Labeled
import javafx.scene.control.ProgressIndicator
import sp.it.pl.util.reactive.Disposer
import sp.it.pl.util.reactive.on
import sp.it.pl.util.reactive.sync
import sp.it.pl.util.reactive.syncTo

/** Provides information about the task and its progress. */
open class TaskInfo<T: Task<*>>: NodeInfo<T> {

    /** Title label or null if none. */
    val title: Labeled?
    /** Message label or null if none. */
    val message: Labeled?
    /** Progress indicator or null if none. */
    val progress: ProgressIndicator?

    protected val disposer = Disposer()

    constructor(title: Labeled?, message: Labeled?, progressIndicator: ProgressIndicator?) {
        this.title = title
        this.message = message
        this.progress = progressIndicator
    }

    override fun setVisible(v: Boolean) {
        title?.isVisible = v
        message?.isVisible = v
        progress?.isVisible = v
    }

    override fun bind(bindable: T) {
        unbind()
        val computeProgress = { it: Number -> if (bindable.state==SCHEDULED || bindable.state==READY) 1.0 else it.toDouble() }
        if (title!=null) bindable.titleProperty() syncTo title.textProperty() on disposer
        if (message!=null) bindable.messageProperty() syncTo message.textProperty() on disposer
        if (progress!=null) bindable.progressProperty() sync { progress.progress = computeProgress(it) } on disposer
    }

    override fun unbind() = disposer()

}